memcache/memcache.go

```diff
@@ -238,13 +278,91 @@ func (c *Client) netTimeout() time.Duration {
 	return DefaultTimeout
 }
 
+func (c *Client) minIdleConnsHeadroom() float64 {
+	if c.MinIdleConnsHeadroom < 0 || c.MinIdleConnsHeadroom >= 1 {
+		return 0
+	}
+
+	return c.MinIdleConnsHeadroom
+}
+
 func (c *Client) maxIdleConns() int {
 	if c.MaxIdleConns > 0 {
 		return c.MaxIdleConns
 	}
 	return DefaultMaxIdleConns
 }
 
+func (c *Client) Close() {
+	close(c.closed)
+}
+
+func (c *Client) releaseIdleConnectionsUntilClosed() {
+	for {
+		select {
+		case <-time.After(releaseIdleConnsCheckFrequency):
+			c.releaseIdleConnections()
+		case <-c.closed:
+			return
+		}
+	}
+}
+
+func (c *Client) releaseIdleConnections() {
+	var toClose []io.Closer
+
+	// Nothing to do if min idle connections headroom is disabled (zero value).
+	minIdleHeadroom := c.minIdleConnsHeadroom()
+	if minIdleHeadroom <= 0 {
+		return
+	}
+
+	// Get the recently used connections threshold, falling back to the default one.
+	recentlyUsedThreshold := c.recentlyUsedConnsThreshold
+	if recentlyUsedThreshold == 0 {
+		recentlyUsedThreshold = defaultRecentlyUsedConnsThreshold
+	}
+
+	c.lk.Lock()
+
+	for addr, freeConnections := range c.freeconn {
+		numIdle := 0
+
+		// Count the number of idle connections. Since the least used connections are at the beginning
+		// of the list, we can stop searching as soon as we find a non-idle connection.
+		for _, freeConn := range freeConnections {
+			if time.Since(freeConn.idleSince) < recentlyUsedThreshold {
+				break
+			}
+
+			numIdle++
+		}
+
+		// Compute the number of connections to close, honoring two properties:
+		// 1. Keep a number of idle connections equal to the configured headroom.
+		// 2. Keep at least 1 idle connection.
+		numRecentlyUsed := len(freeConnections) - numIdle
+		numIdleToKeep := int(math.Max(1, math.Ceil(float64(numRecentlyUsed)*minIdleHeadroom)))
+		numIdleToClose := len(freeConnections) - numRecentlyUsed - numIdleToKeep
```

Comment:
Maybe this is easier to follow? Not a strong opinion:

```suggestion
		numIdleToClose := numIdle - numIdleToKeep
```

Reply:
👍 Done in https://github.com/grafana/gomemcache/pull/10/commits/c873d2f48a2fa56f66e751f71e6a9003f208c702.